cli/azd/docs/external-prompting.md

```diff
@@ -0,0 +1,114 @@
+# External Prompting
+
+## Problem
+
+During operations, `azd` may need to prompt the user for information. For example, during `init` for an Aspire application, we prompt the user to select which services should be exposed to the Internet. The first time `azd provision` is run for an environment, we ask the user to select an subscription and location. In addition, the IaC we provision for an application may have parameters which `azd` needs to prompt the user for.
+
+Today, this prompting happens via the terminal - we have a set of methods that can be called on `input.Console` that allow different types of prompts:
+
+- `Prompt`: Asks the user for a single (freeform) text value.
+- `Select`: Asks the user to pick a single value from a list of options.
+- `MulitSelect`: Asks the user to pick zero or more values from a list of options.
+- `Confirm`: Asks the user to confirm and operation with a given message.
+
+The implementation of this interface uses a go library to provide a terminal experience (using ANSI escape sequences to provide a nice terminal interaction model) with a fallback to raw text input when the user is not connected to a proper terminal.
+
+This is a reasonable experience for users interacting with `azd` via their terminal.  However, `azd` also supports being used within IDEs (today Visual Studio Code, tomorrow Visual Studio as well) and there our terminal based prompting strategy is not ideal. VS Code is forced to run us in an interactive terminal and the user has to interact with `azd` via its terminal interface or specifically craft their calls of `azd` to not prompt the user.  In Visual Studio, AZD is run in a background process, so no terminal interaction is possible.
+
+In both cases it would be ideal if `azd` could delegate the prompting behavior back to the caller.  This document outlines a solution, which provides a way for an external tool to provide a remote service that `azd` interacts with when it needs to prompt the user for information.
+
+## Solution
+
+Similar to our strategy for delegating authentication to an external host, we support delegating prompting to an external host via a special JSON based REST API, hosted over a local HTTP server.  When run, `azd` looks for two special environment variables:
+
+- `AZD_UI_PROMPT_ENDPOINT`
+- `AZD_UI_PROMPT_KEY`
```

Comment:
The choice of communicating through HTTP instead of existing channels: stdout/stdin/stderr is an interesting one. I do like how this is similar to the auth story (although there was a general protocol/established communication which made it seem natural).
 
Anyways, I don't think the tradeoffs are favored in either way. The only minor critique I have is how we think for future scaling. Suppose we have additional communication needed back-and-forth between the CLI and the IDE, will we continue introducing more different endpoint variables? Supposedly we could decide later on "Here's swagger for the full azd-supported communications" and how to set each one.

Reply:
> The choice of communicating through HTTP instead of existing channels: stdout/stdin/stderr is an interesting one. I do like how this is similar to the auth story (although there was a general protocol/established communication which made it seem natural).

I came from a position of "VS Code will be the first consume here, and in that world we are already using the standard io streams connected to the interactive terminal, so we need some out of band transport".

> Anyways, I don't think the tradeoffs are favored in either way. The only minor critique I have is how we think for future scaling. Suppose we have additional communication needed back-and-forth between the CLI and the IDE, will we continue introducing more different endpoint variables? Supposedly we could decide later on "Here's swagger for the full azd-supported communications" and how to set each one.

I could imagine we unify on some common name and then do some form of "discovery" using `OPTIONS` (or perhaps even just `GET`?) on the endpoint to see what services (external auth, external prompting, etc) are present and should be enabled?

The unified API right now is just:

`/token`
`/prompt`

You could imagine that doing `OPTIONS` on `/` would give you back some response with a header like `azd-supported-services: token, prompt` to let you know that you should do both external auth and prompting? Doing an `OPTIONS` on `/prompt` could give you a set of supported API-versions.

Reply:
Also note: my hope is that we design the API in an agnostic enough manner that it's easy to lift and use with our JSON RPC strategy as well.  I can imagine that long term we will want to move VS Code onto this plan.  So hopefully there's a clear separation in the design between the transport and application layer concerns.

Reply:
> You could imagine that doing OPTIONS on / would give you back some response with a header like azd-supported-services: token, prompt to let you know that you should do both external auth and prompting

That's nice. Thank you for entertaining my curiosity!

> Also note: my hope is that we design the API in an agnostic enough manner that it's easy to lift and use with our JSON RPC strategy as well

Agreed. The data contract as well as the general flow-of-control should be reshared.